LinearWeightNorm.lua

```diff
@@ -0,0 +1,220 @@
+    if self.dirty or self.train then
+        if self.norm:dim() == 0 then self.norm:resizeAs(self.g) end
+        if self.scale:dim() == 0 then self.scale:resizeAs(self.g) end
+        if self.weight:dim() == 0 then self.weight:resizeAs(self.v) end
+
+        self.norm:norm(self.v,2,2):add(self.eps)
+        self.scale:copy(self.g):cdiv(self.norm)
+        self.weight:copy(self.v):cmul(self.scale:expandAs(self.v))
+        
+        self.dirty = false
+    end
+end
+
+function LinearWeightNorm:updateOutput(input)
+    self:updateWeightMatrix()
+    
+    if input:dim() == 1 then
+        self.output:resize(self.weight:size(1))
+        if self.bias then self.output:copy(self.bias) else self.output:zero() end      
+        self.output:addmv(1, self.weight, input)
+    elseif input:dim() == 2 then
+        local nframe = input:size(1)
+        local nElement = self.output:nElement()
+        self.output:resize(nframe, self.weight:size(1))
+        if self.output:nElement() ~= nElement then
+            self.output:zero()
+        end
+        updateAddBuffer(self, input)      
+        self.output:addmm(0, self.output, 1, input, self.weight:t())
+        if self.bias then self.output:addr(1, self.addBuffer, self.bias) end
+    else
+        error('input must be vector or matrix')
+    end
+
+    return self.output
+end
+
+function LinearWeightNorm:updateGradInput(input, gradOutput)   
+    if self.gradInput then
+        local nElement = self.gradInput:nElement()
+        self.gradInput:resizeAs(input)
+        if self.gradInput:nElement() ~= nElement then
+            self.gradInput:zero()
+        end
+        if input:dim() == 1 then
+            self.gradInput:addmv(0, 1, self.weight:t(), gradOutput)
+        elseif input:dim() == 2 then
+            self.gradInput:addmm(0, 1, gradOutput, self.weight)
+        end
+
+        return self.gradInput
+    end
+end
+
+function LinearWeightNorm:accGradParameters(input, gradOutput, scale)   
+    scale = scale or 1
+    if input:dim() == 1 then
+        self.gradV:addr(scale, gradOutput, input)
+        if self.bias then self.gradBias:add(scale, gradOutput) end
+    elseif input:dim() == 2 then
+        self.gradV:addmm(scale, gradOutput:t(), input)
+        if self.bias then
+            -- update the size of addBuffer if the input is not the same size as the one we had in last updateGradInput
+            updateAddBuffer(self, input)
+            self.gradBias:addmv(scale, gradOutput:t(), self.addBuffer)
+        end
+    end
+
+    local scale = self.scale:expandAs(self.v)
+    local norm = self.norm:expandAs(self.v)
+
+    self.weight:copy(self.gradV):cmul(self.v):cdiv(norm) -- reuse the weight tensor, avoid unnecessary alloc
+    self.gradG:sum(self.weight,2)
+    
+    self.gradV:cmul(scale)
+
+    self.weight:copy(self.v):cmul(scale):cdiv(norm)
+    self.weight:cmul(self.gradG:expandAs(self.weight))
+
+    self.gradV:add(-1,self.weight)
+
+    self.dirty = true
```

Comment:
The way I understand the `dirty` flag is that if true it indicates that the rows of self.weight are not on the hyper-spheres with respective radiuses  self.g. Setting it to true here seems a bit premature since the module's state only becomes 'dirty' after a parameter update. 
I don't think you  really need this line. For users who manually update the parameters, i.e. without calling `updateParameters`, there is already a safety in place, since when in training mode, the module always behaves as though it's dirty (see line 100).